enterprise_subsidy/apps/transaction/management/commands/write_reversals_from_enterprise_unenrollments.py

```diff
@@ -0,0 +1,209 @@
+            logger.info(
+                f"Unenrollment from course: {enrollment_course_run_key} by user: "
+                f"{enterprise_course_enrollment.get('enterprise_customer_user')} is not refundable."
+            )
+            return False
+
+    def handle(self, *args, **options):
+        """
+        Fetch enterprise enrollment objects that have been unenrolled within the last 24 hours and write Transaction
+        Reversals where appropriate for each unenrollment.
+        """
+        DRY_RUN_LOG_PREFIX = ""
+        if dry_run := options.get('dry_run'):
+            DRY_RUN_LOG_PREFIX = "DRY RUN: "
+            logger.info("Running in dry-run mode. No updates or creates will occur.")
+
+        logger.info(
+            f"{DRY_RUN_LOG_PREFIX}Updating and writing Transaction Reversals from recent Enterprise unenrollments data"
+        )
+        recent_unenrollments = EnterpriseApiClient().fetch_recent_unenrollments()
+        logger.info(
+            f"{DRY_RUN_LOG_PREFIX}Found {len(recent_unenrollments)} recent Enterprise unenrollments"
+        )
+
+        reversals_processed = 0
+        enterprise_catalog_api_client = EnterpriseCatalogApiClient()
+        fetched_content_metadata = {}
+        for unenrollment in recent_unenrollments:
+            fulfillment_uuid = unenrollment.get("uuid")
+            enterprise_course_enrollment = unenrollment.get("enterprise_course_enrollment")
+            enrollment_course_run_key = enterprise_course_enrollment.get("course_id")
+            enrollment_unenrolled_at = enterprise_course_enrollment.get("unenrolled_at")
+            try:
+                # Look for a transaction related to the unenrollment
+                # This will throw a Transaction.DoesNotExist exception if no transaction is found
+                related_transaction = Transaction.objects.get(uuid=unenrollment.get('transaction_id'))
+                idempotency_key = generate_transaction_reversal_idempotency_key(
+                    fulfillment_uuid,
+                    enrollment_unenrolled_at
+                )
+                # This ORM query will throw a Reversal.DoesNotExist exception if no reversal is found,
+                # which means we can move forward with the reversal generation
+                existing_reversal = Reversal.objects.get(
+                    transaction=related_transaction,
+                    idempotency_key=idempotency_key,
+                )
+            except Transaction.DoesNotExist:
+                logger.info(
+                    f"{DRY_RUN_LOG_PREFIX}No Subsidy Transaction found for enterprise fulfillment: {fulfillment_uuid}"
+                )
+                continue
+            except Reversal.DoesNotExist:
+                logger.info(
+                    f"{DRY_RUN_LOG_PREFIX}No existing Reversal found for enterprise fulfillment: {fulfillment_uuid}. "
+                    f"Writing Reversal for Transaction: {related_transaction}."
+                )
```

Comment:
nit: I would maybe re-organize this code to not do so much of the real work inside your exception handling blocks.  That would allow to early-continue through your loop when you find that there _is_ already a reversal for the transaction.  The `Transaction.get_reversal()` method will give you a reversal or none, safely.

Reply:
@iloveagent57 My only concern with `get_reversal` that I see is the lack of an idempotency key consideration. Is that something we're ok with dropping?

Reply:
There was a time when we were considering the possibility of multiple reversals for one transaction, but that is not the case anymore.  It is no longer clear to me the utility of an idempotency key for finding existing reversals since the reversal you're looking for can always be precisely located by their linked transaction UUID.  In other words, I'm fine with just doing a transaction-based lookup on reversals.

Reply:
It still begs the question, why have an idempotency key field at all for reversals!  Rocking the boat a little.

Reply:
to check for a previous reversal, we shouldn't use the idemptoency key - any prior reversal should count as "this transaction has been reversed". when applying our reversal, we should include our idemptotency key. this is good practice. it prevents multiple copies of the same code from stomping on each other. makes it safer to-rerun.

Reply:
idempotency keys are important for callers to supply when taking an action that alters the state of the ledger. idempotency keys are only important on filtering if you really, truely only want to know if this specific code path has already run. usually knowing that _any code path has already taken this action_ is the more correct thing to ask (ie "has this transaction been reversed?" means any code path, "do you have a reversal with this idempotency key?" means this one code path)